enhancements/machine-api/aws-imds-v2-support.md

```diff
@@ -0,0 +1,270 @@
+At this point, IMDSv2 enablement might be done manually via AWS console, but providing such settings on Machine API is a quite loud request from our end-users.
+
+### Goals
+
+- provide mechanisms to enable/enforce IMDSv2 for newly created machines via Machine API
+- provide mechanisms to enable/enforce IMDSv2 for master machines during installation procedure for newly created clusters
+
+### Non-Goals
+
+- perform automated migration between using IMDSv1 and IMDSv2 for existing infrastructure
+
+## Proposal
+
+- Add new options into install-config for the OCP installation program which will be passed as metadata service settings during virtual machines creation
+- Extend [AWSMachineProviderConfig](https://github.com/openshift/api/blob/1a6fa2913810101176a1d776f899fc4781b3fa50/machine/v1beta1/types_awsprovider.go#L12) by adding metadata service parameters
+- Pass metadata service parameters to AWS during new machines creation
+
+### User Stories
+
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage during an OCP cluster deployment procedure.
+- As an OpenShift cluster administrator, I would like to enforce IMDSv2 usage for newly created machines in my cluster via MachineSets configuration.
+- As an OpenShift cluster administrator, I would like to change IMDS interaction settings for newly created machines in my cluster.
+
+### API Extensions
+
+#### AWSMachineProviderConfig changes
+
+MachineAPI design mostly replicates parameters which [AWS API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html) consumes.
+
+```go
+type AWSMachineProviderConfig struct {
+    // Existing fields will not be modified
+    ...
+    // MetadataServiceOptions allows users to configure instance metadata service interaction options.
+    // If nothing specified, default AWS IMDS settings will be applied.
+    // https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_InstanceMetadataOptionsRequest.html
+    // +optional
+    MetadataServiceOptions *MetadataServiceOptions `json:"metadataServiceOptions,omitempty"`
+}
+```
+
+```go
+
+type InstanceMetadataHTTPTokens string
+
+const (
+    // InstanceMetadataHTTPTokensRequired enforces sending of a signed token header with any instance metadata retrieval (GET) requests.
+    // Enforces IMDSv2 usage.
+    InstanceMetadataHTTPTokensRequired = "required"
+    // InstanceMetadataHTTPTokensOptional allows IMDSv1 usage along with IMDSv2
+    InstanceMetadataHTTPTokensOptional = "optional"
+)
+
+// MetadataServiceOptions defines the options available to a user when configuring
+// Instance Metadata Service (IMDS) Options.
+type MetadataServiceOptions struct {
+    // HTTPTokens determines whether or not the host requires the use of authentication when interacting with the metadata service.
```

Comment:
The name of this still I think needs work, should it not contain Auth in it for example, `HTTPAuthTokens`? I'm just thinking that as a user, the name of the field looks very opaque

```
metadataServiceOptions:
  httpAuthTokens: required
```
vs
```
metadataServiceOptions:
  httpTokens: required
```
TBH, I think I might even go as far as to say we just call it authentication, because that's what we are enforcing here right?
```
metadataServiceOptions:
  authentication: required
```

Reply:
```
When token usage is set to required, requests without a valid token or with an expired token receive a 401 - Unauthorized HTTP error code.
```

This is from the AWS docs, so I think we can assert that the token is a form of authentication.

> It's AWS specific parameter existing only in AWS context. I'm strongly against new naming invention here.

I totally get where you're coming from, however, imagine as an end user you look at the structure of this
```
metadataServiceOptions:
  httpTokens: required
```
Would you know what that is doing without having prior knowledge of the IMDS configuration on AWS? My argument is merely that we can make it much more obvious to an end user by choosing a more descriptive name.

Let's gather the opinions of a few others on this and see if we can come up with a conclusion as a group

Reply:
> Would you know what that is doing without having prior knowledge of the IMDS configuration on AWS?

I expect end user will face AWS documentation at some point. And close naming will help to build an obvious connection between which we and they offering.

Basically i trying to follow the principle of least astonishment, and to not introduce additional obscurity layer without necessity.

Also, i can supply my position a bit by reffering to [terraform](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#metadata-options) interface for this functionality which is using `http_tokens` name for this knob.

Reply:
i like the notion of staying with `HTTPTokens` as that does stay close to the other sources that a user will be interacting with. i can understand @JoelSpeed 's point here about making it more clear and not repeating things from other APIs just because they have the same name, it seems like this may be a case where staying with the original wouldn't be so bad (least astonishment).

if we do decide to change this to something more verbose, we need to make sure that the "HTTPTokens" is referenced in the comments so that a user will see what the original value in the AWS API is.

so, i guess, i'm +1 for keeping it as is.